plugins/source/azure/client/helpers.go

```diff
@@ -22,3 +23,28 @@ func ParseResourceGroup(resourceID string) (string, error) {
 	}
 	return match[2], nil
 }
+
+type syncData struct {
+	data any
+	once *sync.Once
+}
+
+func loadOrStore(m *sync.Map, key string, f func() (any, error)) (any, error) {
+	temp, _ := m.LoadOrStore(key, &syncData{
+		data: nil,
+		once: &sync.Once{},
+	})
+	d := temp.(*syncData)
+	var err error
+	if d.data == nil {
+		d.once.Do(func() {
+			d.data, err = f()
+			if err != nil {
+				d.data = nil
+				d.once = &sync.Once{}
+			}
+			m.Store(key, d)
+		})
+	}
+	return d.data, err
```

Comment:
Nice solution, but I think there might be a weird edge case here... if `f` returns an error. In this case the function to execute `d.once.Do` will have an error and return it, but all the other callers that were waiting on the `d.once.Do` to finish will return with nil data and no error, which is not quite right. Maybe there needs to be another check at the end of the function to return an error if `d.data` is still nil, or something along those lines.

Reply:
Should be fixed with [`733b79a` (#10554)](https://github.com/cloudquery/cloudquery/pull/10554/commits/733b79a7facafea5222cc34f563070419ce763b5)